### PR TITLE
MPA-5562 Implement gateway using RSocket protocol over WebSocket

### PR DESCRIPTION
**Pull request by dmytro-lazebnyi**

### CODE REVIEW

File: RSocketWebSocket/src/main/java/io/scalecube/gateway/rsocket/core/RSocketGatewayMessageCodec.java [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+
+    final ByteBuf metadata = payload.sliceMetadata();
+
+    if (metadata.isReadable()) {
+      try (ByteBufInputStream stream = new ByteBufInputStream(metadata.slice())) {
+        final Map<String, String> metadataMap = jacksonCodec.decode(stream);
+        builder.qualifier(metadataMap.get(QUALIFIER_FIELD));
+      } catch (Throwable ex) {
+        LOGGER.error("Failed to decode metadata: {}, cause: {}", metadata.toString(UTF_8), ex);
+        throw new GatewayException("Failed to decode metadata", ex);
```
Comment (artem-v) [inline RSocketWebSocket/src/main/java/io/scalecube/gateway/rsocket/core/RSocketGatewayMessageCodec.java]: Use instead `MessageCodecException`

File: RSocketWebSocket/src/main/java/io/scalecube/gateway/rsocket/core/RSocketGatewayMessageCodec.java [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+  public Payload encode(GatewayMessage gatewayMessage) {
+    ByteBuf metadata = ByteBufAllocator.DEFAULT.buffer();
+    try {
+      Map<String, String> metadataMap = new HashMap<>();
+      metadataMap.put(QUALIFIER_FIELD, gatewayMessage.qualifier());
+      jacksonCodec.encode(new ByteBufOutputStream(metadata), metadataMap);
+    } catch (Throwable ex) {
+      ReferenceCountUtil.safeRelease(metadata);
+      LOGGER.error("Failed to encode metadata: {}, cause: {}", gatewayMessage, ex);
+      throw new GatewayException("Failed to encode metadata on message q=" + gatewayMessage.qualifier(), ex);
```
Comment (artem-v) [inline RSocketWebSocket/src/main/java/io/scalecube/gateway/rsocket/core/RSocketGatewayMessageCodec.java]: Use `MessageCodecExceptgion` instead

File: RSocketWebSocket/src/main/java/io/scalecube/gateway/rsocket/core/GatewayException.java [hunk 3]
```diff
@@ -0,0 +1,13 @@
+package io.scalecube.gateway.rsocket.core;
+
+public class GatewayException extends RuntimeException {
```
Comment (artem-v) [inline RSocketWebSocket/src/main/java/io/scalecube/gateway/rsocket/core/GatewayException.java]: Remove it plz

File: pom.xml [hunk 4]
```diff
@@ -93,14 +94,34 @@
     <module>ExamplesRunner</module>
     <module>Tests</module>
     <module>Benchmarks</module>
+    <module>RSocketWebSocket</module>
```
Comment (artem-v) [inline pom.xml]: Plz rename to `RSocketWebsocket`

File: pom.xml [hunk 5]
```diff
@@ -93,14 +94,34 @@
...
+    <module>RSocketWebSocketRunner</module>
```
Comment (artem-v) [inline pom.xml]: Plz rename to `RSocketWebsocketRunner`

File: Tests/src/test/java/io/scalecube/gateway/rsocket/RSocketGatewayResource.java [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+
+public class RSocketGatewayResource extends ExternalResource {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(RSocketGatewayResource.class);
+
+
+  private RSocketWebsocketGateway gateway;
+  private RSocketGatewayClient client;
+
+  public RSocketWebsocketGateway startGateway(Microservices microservices) {
```
Comment (ronenhamias) [inline Tests/src/test/java/io/scalecube/gateway/rsocket/RSocketGatewayResource.java]: please conform to start method keeping same api as in websocket gw

File: RSocketWebsocketRunner/src/main/java/io/scalecube/gateway/rsocket/websocket/RSocketWebSocketGatewayRunner.java [hunk 7]
```diff
@@ -0,0 +1,24 @@
+package io.scalecube.gateway.rsocket.websocket;
+
+import io.scalecube.services.Microservices;
+import io.scalecube.transport.Address;
+
+public class RSocketWebSocketGatewayRunner {
+
+  private static final Address SEED_ADDRESS = Address.from("localhost:4801");
```
Comment (aharonha) [inline RSocketWebsocketRunner/src/main/java/io/scalecube/gateway/rsocket/websocket/RSocketWebSocketGatewayRunner.java]: Hardcoded?

Comment (dmytro-lazebnyi) [reply RSocketWebsocketRunner/src/main/java/io/scalecube/gateway/rsocket/websocket/RSocketWebSocketGatewayRunner.java]: Fixed.

**Pull request merged**